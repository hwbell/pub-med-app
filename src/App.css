@import url('https://fonts.googleapis.com/css?family=Quicksand');
* {
  font-family: Quicksand;
}
body {
  background-attachment: fixed;
  background:url(./assets/background-image.jpg) fixed center;
  background-size:100vw auto;
  min-height: 100vh;
}
* > p {
  color: whitesmoke;
  font-size: 16px;
}
* > a {
  text-decoration: none;
}


hr {
  /* border: 0; */
  width: 100%;               
  background-color: grey;
  height: 0.3px;
  margin-bottom: 10px;
}

/* to remove spacing between bootstrap cols - https://stackoverflow.com/questions/40267056/remove-spacing-between-cols */
.padding-0 {
  padding-right:0;
  padding-left:0;
}

/*  use these! */
.center-all-col {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.heading, .glass, .outline, .article-result {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
}

.article-result {
  width: 100%; 
  margin: 2px;
  border-radius: 4px;
  background-color: rgba(0, 0, 0, 0.24);
  padding: 10px;
}

.page {
  max-width: 800px;
  margin: 0px auto 80px auto;
  padding-top: 120px;
}

/* for any glass background container, add this to the parent div */
.glass {
  position:relative;
  background-color: rgba(11, 0, 47, 0.24);
}
.glass:before{
  content:'';
  position:absolute;
  z-index:-1;
  top:0;
  left:0;
  bottom:0;
  right:0;
  background:url(./assets/background-image.jpg) fixed center;
  filter:blur(10px);
  background-size:100vw auto;
  /* box-shadow:inset 0 0 0 50vw rgba(255,255,255,0.2) */
}

/* for any outlined container, add this to the parent div */
.outline {
  border: 2px solid rgba(255, 255, 255, 0.226);
  width: 92%;
  margin: 15px;
}

/* this will be the universal border-radius */
.glass, .outline {
  border-radius: 10px;
}

/* to make bootstrap columns equal height */
.row-eq-height{
  margin-top: 4px;
  display: flex;
  flex-wrap: wrap;
}

/* some common text / button classes */
.title, .subtitle {
  color: whitesmoke;
  padding: 10px 25px;
}
.title {
  font-size: 40px;
}
.subtitle {
  padding-top: 0px;
  font-size: 24px;
}
.text, .paragraph {
  font-size: 14px;
}
.text {
  padding: 0px 25px;
}
.paragraph {
  padding: 10px 20px;
}
.article-title {
  padding: 0px 20px;
}
.article-title {
  font-size: 16px;
}
.collection-title {
  padding: 10px 20px 0px 20px;
  font-size: 22px;
}

/* buttons */
.block-button, .article-button, .add, .warn {
  border: none;
}
.block-button {
  align-self: flex-start;
  margin: 10px 25px;
}
.article-button {
  margin: 4px;
}
.warn {
  background-color: rgb(194, 96, 85);
}
.warn:hover {
  background-color: rgb(150, 70, 63);
}
.add {
  background-color: rgb(26, 98, 214);
}
.add:hover {
  background-color: rgb(0, 74, 194);
}


/* the nav component */
.navigator {
  background-color: white;
  /* position: absolute; */
  top: 0px;
  left: 0px;
}
#logo {
  font-size: 32px;
  color: rgba(0, 60, 255, 0.94)
}
.nav-link {
  /* color: rgb(253, 233, 255); */
  transition: all 0.3s ease;
}
.nav-link:hover {
  color: rgb(161, 161, 161);
}


/* the search page */
.search-input {
  width: 60%;
  margin: 15px;
}
input {
  margin: 15px;
  text-align: center;
}


/* the collection page */

/* the div holding the react-pdf PDFViewer */
.pdf-holder {
  width: 100%;
  height: 600px;
}
/* the PDFViewer itself */
.pdf-viewer {
  /* border: 1px solid red; */
  width: 100%;
  height: 100%;
}


/* reactstrap modals */
/* this setting prevents the jumpy shift when the scrollbar disappears on modal popup */
body.modal-open {
  overflow: visible;
  padding-right: 0 !important;
}
.modal-xl {
 /* margin: 0px; */
 padding: 20px;
 max-width: 750px;
}