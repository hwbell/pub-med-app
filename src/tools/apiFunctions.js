const fetch = require('node-fetch')

// query and return based on date / citations
const getArticles = async (query, sortParam) => {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;
  
  let headers = {}
  headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let articles = await fetch(searchUrl, {
    method: 'POST',
    headers
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return articles;
}

// parses the result of a search query to a simple array of the article titles
const parseSearchToTitlesArray = (searchResults) => {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })
  return titlesArray;
}

// let results = getArticles('cancer').then((response) => {
//   console.log(response.resultList.result.slice(0,2))
// })

module.exports = {
  getArticles,
  parseSearchToTitlesArray,
}
