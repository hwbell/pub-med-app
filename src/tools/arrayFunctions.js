// this function will remove an article from a specific collection in the collections array
const removeArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles = collection.articles.filter(item => item.id !== article.id)
    }

  });

  return collections;
}

// this function will add an article to a specific collection
const addArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles.push(article);
    }

  });

  return collections;
}

module.exports = {
  removeArticle,
  addArticle
}