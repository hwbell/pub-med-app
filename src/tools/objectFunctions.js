
// this comibines two objects for the keys specified, giving priority to obj1
const combineObjects = (obj1, obj2, keys) => {

  if(!obj1 || !obj2) {
    return console.log('object argument missing')
  }
  if(!keys || !keys.length) {
    return console.log('keys argument missing')
  }

  let combined = {};

  keys.forEach((key) => {
    if (obj1[key]) {
      combined[key] = obj1[key];
    } else if (obj2[key]){
      combined[key] = obj2[key];
    } else {
      combined[key] = '';
    }
  })

  return combined;
}

module.exports = {
  combineObjects
}