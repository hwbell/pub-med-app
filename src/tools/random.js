let list = [];
for (let i = 0; i < 50; i++) {
  list.push({
    title: `Article ${i + 1}`,
    pmid: `PMID10000${i}`
  })
}

console.log(list)