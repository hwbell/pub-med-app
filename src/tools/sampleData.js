
// set some sample data for tests

export function getSampleCollections() {
  let collections = [];
  for (let i = 0; i < 10; i++) {
    collections.push({
      name: `collection ${i + 1}`,
      articles: [
        {
          name: `article ${i + 1}`,
          name: `article ${i + 2}`,
          name: `article ${i + 3}`,
        }
      ]
    })
  }
  return collections;
}