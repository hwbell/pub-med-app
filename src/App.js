import React, { Component } from 'react';
import 'bootstrap/dist/css/bootstrap.css';
import './App.scss';

// components
import HomePage from './components/pages/HomePage';
import AboutPage from './components/pages/AboutPage';
import SearchPage from './components/pages/SearchPage'
import CollectionPage from './components/pages/CollectionPage';
import ThreadPage from './components/pages/ThreadPage';
import { Dropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap';
import Media from "react-media";

// pose animation
import posed, { PoseGroup } from 'react-pose';
import styled from "styled-components";

// routing
import { Switch, Link } from 'react-router-dom';
import ProfilePage from './components/pages/ProfilePage';

// tools
import { getUserCollections } from './tools/serverFunctions';
import { addArticle, removeArticle } from './tools/arrayFunctions';

// more routing .... dev console warning said to change it to this instead of import?
const Router = require("react-router-dom").BrowserRouter;
const Route = require("react-router-dom").Route;

const RoutesContainer = posed.div({
  enter: {
    transition: {
      ease: 'easeIn', duration: 300
    },
    opacity: 1,
    scale: 1
  },
  exit: {
    delay: 300,
    transition: {
      ease: 'easeOut', duration: 300
    },
    opacity: 0,
    scale: 0.01
  }
});

const links = [
  {
    title: 'About',
    link: '/about'
  },
  {
    title: 'Search',
    link: '/search'
  },
  {
    title: 'Collections',
    link: '/collections'
  },
  {
    title: 'Threads',
    link: '/threads'
  },
]

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      collections: [],
      serverThreads: null,
      user: null,
      userCollections: [],
      dropdownOpen: false
    };

    this.toggle = this.toggle.bind(this);
    this.renderNavigator = this.renderNavigator.bind(this);
    this.renderDropdown = this.renderDropdown.bind(this);
    // these functions are for collections that are only in the App state, not yet stored
    // on the backend
    this.createNewCollection = this.createNewCollection.bind(this);
    this.modifyCollection = this.modifyCollection.bind(this);
    this.deleteCollection = this.deleteCollection.bind(this);

    // these actually communicate with our server
    this.registerSignIn = this.registerSignIn.bind(this);
    this.registerSignOut = this.registerSignOut.bind(this);
    this.refreshUserCollections = this.refreshUserCollections.bind(this);
    this.refreshUserThreads = this.refreshUserThreads.bind(this);
    this.refreshUser = this.refreshUser.bind(this);

    this.refreshServerThreads = this.refreshServerThreads.bind(this);
  }

  componentDidMount() {
    // sign in on startup if there is a token in localStorage
    if (localStorage.getItem('user')) {
      let user = JSON.parse(localStorage.getItem('user'));
      return this.registerSignIn(user);
    }
  }

  createNewCollection(collection, callback) {
    console.log('creating new collection');

    let collections = this.state.collections;
    collections.push(collection);

    this.setState({
      collections
    }, () => callback());
  }

  deleteCollection(collectionToDelete, callback) {
    let name = collectionToDelete.name;
    console.log(name)
    let collections = this.state.collections.filter(collection => collection.name !== name);

    this.setState({
      collections
    }, () => {
      callback();
    })

  }

  // this will be called:
  //   1. when adding an article from the search page
  //   2. when removing an article from the collection page (only state collections, not server collections)
  modifyCollection(article, collectionName, change, callback) {

    // if no collectionName / article, return
    if (!article || !collectionName) {
      return;
    }

    // to handle an object with 'name' key as well
    if (collectionName.name) {
      collectionName = collectionName.name
    }

    let verb = change > 0 ? 'Adding' : 'Removing';
    console.log(`${verb} article ${article.id} in user's ${collectionName} collection`)

    let collections = this.state.collections;

    if (change > 0) {
      collections = addArticle(collections, collectionName, article)
    } else {
      collections = removeArticle(collections, collectionName, article)
    }

    console.log(collections)

    if (callback) {
      this.setState({
        collections
      }, () => callback());
    } else {
      this.setState({ collections });
    }
  }

  // fire this function 
  // 1. when a user signs in from the user page 
  // 2. Upon startup if there is a localStorage user + token

  // we'll save the user to the state / localStorage, as well as retrieve the user's collections
  // and save that too
  registerSignIn(user) {

    this.setState({ user });

    let token = JSON.parse(localStorage.getItem('token'));
    let headers = {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
      'Content-Type': 'application/json',
    }

    // we actually get the collections as well as the profile here, maybe change the name
    return getUserCollections(headers)
      .then((response) => {

        // attach the collections to the user object
        let profile = response.user;
        if (!user) {
          return;
        }
        profile.collections = response.collections;
        profile.threads = response.threads;


        // save to state
        this.setState({
          user: profile
        }, () => {
          console.log(this.state.user)
        });
        // save info to local storage
        localStorage.setItem(`user`, JSON.stringify(profile));

      }).catch((e) => {
        console.log(e)
        this.setState({
          error: e
        })
      })
  }

  // this will also be called from the Profile component
  registerSignOut() {
    this.setState({
      collections: [],
      user: null,
      userCollections: []
    });
    // wipe the user from localStorage to prevent sign in on startup
    localStorage.removeItem('user');
    localStorage.removeItem('token');
  }

  // this function is triggered from the CollectionPage when a user changes their collections
  // on the server. The updated user collections are sent from the server upon update and sent 
  // back to App to keep App as the root source of the current user data
  refreshUserCollections(response) {
    // all we have to do is set the response as the user's collections

    let { user } = this.state;
    user.collections = response;
    this.setState({
      user
    })
  }

  refreshUserThreads(response) {
    // all we have to do is set the response as the user's threads

    let { user } = this.state;
    user.threads = response;
    this.setState({
      user
    }, () => {
      console.log(user)
    })

  }

  // this function is triggered from the ProfileForm in ProfilePage when a user changes their info
  // The updated user is sent from the server upon update and sent 
  // back to App to keep App as the root source of the current user data
  refreshUser(response) {

    // all we have to do is set the response as the user
    // but remember to re-attach the user's collections to the user object
    response.collections = this.state.user.collections;
    response.threads = this.state.user.threads;

    this.setState({
      user: response
    }, () => {
      // save to local storage
      localStorage.setItem(`user`, JSON.stringify(response));
    });

  }

  refreshServerThreads(response) {
    this.setState({
      serverThreads: response
    }, () => {
      console.log(this.state.serverThreads)
    })
  }

  renderNavigator() {
    return (
      <div className="fixed-top navigator" id="full-nav" style={styles.nav}>

        <Link to="/" id="logo" className="nav-link">PubMed</Link>

        <div className="row links-holder">
          {links.map((link, i) => {
            return (
              <Link className="nav-link" style={styles.link} key={i} to={link.link}>{link.title}</Link>
            )
          })}

          <Link className="nav-link" style={styles.link} to="/profile/">
            <i className="fas fa-user-cog"></i>
          </Link>
        </div>


      </div>
    )
  }

  renderDropdown() {
    return (
      <div className="fixed-top navigator" id="dropdown" style={styles.nav}>

        <Link to="/" id="logo" className="nav-link">PubMed</Link>

        <Dropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
          <DropdownToggle color="primary" caret>
            go to
        </DropdownToggle>
          <DropdownMenu>
            {links.map((link, i) => {
              return (
                <Link className="nav-link" style={styles.link} key={i} to={link.link}>{link.title}</Link>
              )
            })}

            <Link className="nav-link" style={styles.link} to="/profile/">
              <i className="fas fa-user-cog"></i>
            </Link>
          </DropdownMenu>
        </Dropdown>


      </div>
    )
  }

  toggle() {
    this.setState(state => ({ dropdownOpen: !state.dropdownOpen }));
  }

  render() {

    return (
      <Router>

        {/* Try and put this Navigator in its own component if necessary? Its not resuasable so maybe
          not. There was an issue with the router(I think just w. enzyme tests) so it is here for now */}

        <Media query="(max-width: 649px)">
          {matches =>
            matches ? (
              this.renderDropdown()
            ) : (
                this.renderNavigator()
              )
          }
        </Media>

        <div id="App">

          <Route render={({ location }) => (

            // pose is awesome 
            <PoseGroup>

              <RoutesContainer key={location.pathname}>

                <Switch location={location}>
                  <Route exact path="/" component={HomePage} />
                  <Route path="/about/" component={AboutPage} />
                  <Route path="/search/" render={() =>
                    <SearchPage
                      collections={this.state.collections}
                      createNewCollection={this.createNewCollection}
                      modifyCollection={this.modifyCollection}
                      user={this.state.user} />
                  } />
                  <Route path="/collections/" render={() =>
                    <CollectionPage
                      collections={this.state.collections}
                      modifyCollection={this.modifyCollection}
                      deleteCollection={this.deleteCollection}
                      refreshUserCollections={this.refreshUserCollections}
                      user={this.state.user}
                    />
                  } />
                  <Route path="/threads/" render={() =>
                    <ThreadPage
                      serverThreads={this.state.serverThreads}
                      user={this.state.user}
                      refreshUserThreads={this.refreshUserThreads}
                      refreshServerThreads={this.refreshServerThreads}
                    />

                  } />
                  <Route path="/profile/" render={() =>
                    <ProfilePage
                      registerSignIn={this.registerSignIn}
                      registerSignOut={this.registerSignOut}
                      refreshUser={this.refreshUser}
                      user={this.state.user}
                    />

                  } />
                </Switch>

              </RoutesContainer>

            </PoseGroup>

          )} />

        </div>
      </Router>
    );
  }
}

export default App;

// export default App;

const styles = {
  nav: {
    padding: '10px',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center'
  },
  link: {
    fontSize: ' 18px'
  }
}