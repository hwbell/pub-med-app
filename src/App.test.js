import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// tools
import { getSampleCollections } from './tools/sampleData';

// mock the server functions
jest.mock('./tools/serverFunctions.js');

// stubs
const renderDropDownStub = jest.fn();
const renderNavigatorStub = jest.fn();

const resizeWindow = (x, y) => {
  window.innerWidth = x;
  window.innerHeight = y;
  window.dispatchEvent(new Event('resize'));
}

describe('App', () => {

  let user, wrapper;
  beforeEach(() => {

    user = {
      name: 'Mark',
      email: 'mark@example.com',
      password: 'password'
    }

    wrapper = shallow(<App />);
    wrapper.instance().registerSignIn = jest.fn();
  })


  it('renders without crashing', async () => {
    wrapper.update();
  });

  it('renders correctly', async () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should not have a user to start with', async () => {
    expect(wrapper.state().user).toBe(null)
  });

  it('contains the correct elements', () => {
    expect(wrapper.find('#App').length).toBe(1)
  })

  it('should try to sign in user if there is a user in local storage', async () => {

    jest.spyOn(wrapper.instance(), 'registerSignIn');

    // with no user it should not attempt
    // there should be no user at the start
    expect(localStorage.getItem('user')).not.toBeTruthy();

    // without user => false
    await wrapper.instance().componentDidMount();
    wrapper.update();
    expect(wrapper.instance().registerSignIn).not.toHaveBeenCalled();

    // set a user
    localStorage.setItem('user', JSON.stringify(user));

    // componentDidMount is where the decision is made whether or not to sign in
    await wrapper.instance().componentDidMount();
    wrapper.update();

    // with user => true
    expect(wrapper.instance().registerSignIn).toHaveBeenCalled();

  })

  it('should render full navigator menu when screen width > 649px', () => {
    const wrapper = mount(<App />);

    resizeWindow(1000, 800);

    wrapper.update();

    // there should always only be one navigator
    expect(wrapper.find('.navigator').length).toBe(1);

    expect(wrapper.find('#full-nav').length).toBe(1);


  })

  // it('should render dropdown menu when screen width < 649px', () => {
  //   const wrapper = mount(<App />);

  //   resizeWindow(600, 400);

  //   wrapper.update();

  //   // there should always only be one navigator
  //   expect(wrapper.find('.navigator').length).toBe(1);

  //   expect(wrapper.find('#dropdown').length).toBe(1);


  // })


  it('should set state correctly when refreshUserCollections() is fired', () => {
    expect(wrapper.state().user).toMatchObject(user);

    // can pass an array or object to this function, to account for different responses
    let newCollections = getSampleCollections();
    // pass an array, which will replace the user collections
    wrapper.instance().refreshUserCollections(newCollections);
    wrapper.update();
    expect(wrapper.state().user.collections).toMatchObject(newCollections);

    // pass an new colection, which will be added to the user collections
    let newObj = JSON.parse(JSON.stringify(newCollections[0]));
    newObj._id = '3242';
    newObj.owner = 'user3242';

    wrapper.instance().refreshUserCollections(newObj);
    wrapper.update();
    // length increased by one, and the new collection is the first in the list
    expect(wrapper.state().user.collections.length).toBe(11);
    expect(wrapper.state().user.collections[0]).toMatchObject(newObj);

    // pass an updated object, which will replace the old object
    newObj.name = 'new name';
    newObj.articles.push({
      name: 'article 3242', 
    });
    console.log(newObj)

    wrapper.instance().refreshUserCollections(newObj);
    wrapper.update();
    // length stays the same, and the new collection is the first in the list
    expect(wrapper.state().user.collections.length).toBe(11);
    expect(wrapper.state().user.collections[0]).toMatchObject(newObj);
  })

})


