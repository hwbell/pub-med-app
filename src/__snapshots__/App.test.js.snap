// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="fixed-top navigator"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": "10px",
      }
    }
  >
    <div
      className="row"
    >
      <a
        className="nav-link"
        href="/"
        id="logo"
        onClick={[Function]}
      >
        PubMed
      </a>
    </div>
    <div
      className="row links-holder"
    >
      <a
        className="nav-link"
        href="/about"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "22px",
          }
        }
      >
        About
      </a>
      <a
        className="nav-link"
        href="/search"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "22px",
          }
        }
      >
        Search
      </a>
      <a
        className="nav-link"
        href="/collections"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "22px",
          }
        }
      >
        Collections
      </a>
      <a
        className="nav-link"
        href="/profile"
        onClick={[Function]}
        style={
          Object {
            "fontSize": "22px",
          }
        }
      >
        <i
          className="far fa-user"
        />
      </a>
    </div>
  </div>,
  <div
    id="App"
  >
    <div>
      <div
        className="page"
      >
        <div
          className="col glass"
        >
          <h2
            className="title"
          >
            PubMed Central
          </h2>
          <h4
            className="subtitle"
          >
            making research accessible
          </h4>
        </div>
        <div
          className="row no-gutters row-eq-height"
        >
          <div
            className="col col-12 col-md-8"
          >
            <div
              className="glass"
              style={
                Object {
                  "height": "100%",
                  "margin": "0px 0px 0px 4px ",
                }
              }
            >
              <h4
                className="subtitle"
                style={
                  Object {
                    "alignSelf": "flex-start",
                    "margin": "10px",
                  }
                }
              >
                About PMC
              </h4>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                PubMed Central (PMC) is a free digital repository that archives 
  publicly accessible full-text scholarly articles that have been published within the 
  biomedical and life sciences journal literature. As one of the major research 
  databases within the suite of resources that have been developed by the National 
  Center for Biotechnology Information (NCBI), PubMed Central is much more than just 
  a document repository. Launched in February 2000, the repository has grown rapidly 
  as the NIH Public Access Policy is designed to make all research funded by the 
  National Institutes of Health (NIH) freely accessible to anyone, and, in addition, 
  many publishers are working cooperatively with the NIH to provide free access to 
  their works.
              </p>
              <button
                aria-label={null}
                className="block-button btn btn-secondary btn-sm"
                onClick={[Function]}
              >
                read more
              </button>
            </div>
          </div>
          <div
            className="col col-12 col-md-4"
          >
            <div
              className="glass"
              style={
                Object {
                  "height": "100%",
                  "margin": "0px 0px 0px 4px ",
                }
              }
            >
              <h4
                className="subtitle"
                style={
                  Object {
                    "alignSelf": "flex-start",
                    "margin": "10px",
                  }
                }
              >
                Resources
              </h4>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                abstracts &gt; 34 million
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                full text articles &gt; 5 million
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                patents &gt; 4 million
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                preprints &gt; 74,000
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                agricola records &gt; 700,000
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                NIH clinical guidelines &gt; 800
              </p>
              <button
                aria-label={null}
                className="block-button btn btn-secondary btn-sm"
                onClick={[Function]}
              >
                find a resource
              </button>
            </div>
          </div>
        </div>
        <div
          className="row no-gutters row-eq-height"
        >
          <div
            className="col-12 col-md-4"
          >
            <div
              className="glass"
              style={
                Object {
                  "height": "100%",
                  "margin": "0px 0px 0px 4px ",
                }
              }
            >
              <h4
                className="subtitle"
                style={
                  Object {
                    "alignSelf": "flex-start",
                    "margin": "10px",
                  }
                }
              >
                Journals
              </h4>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Nature
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Science
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Cardiology and Therapy
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Anatomy and Cell Biology
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Journal of Applied Physics
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                PLoS Genetics
              </p>
              <p
                className="text"
                style={
                  Object {
                    "alignSelf": "flex-start",
                  }
                }
              >
                Frontiers in Chemistry
              </p>
              <button
                aria-label={null}
                className="block-button btn btn-secondary btn-sm"
                onClick={[Function]}
              >
                search literature
              </button>
            </div>
          </div>
          <div
            className="col-12 col-md-8"
          >
            <div
              className="glass"
              style={
                Object {
                  "height": "100%",
                  "margin": "0px 0px 0px 4px ",
                }
              }
            >
              <h4
                className="subtitle"
                style={
                  Object {
                    "alignSelf": "flex-start",
                    "margin": "10px",
                  }
                }
              >
                New Publications
              </h4>
              <button
                aria-label={null}
                className="block-button btn btn-secondary btn-sm"
                onClick={[Function]}
              >
                more articles
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
